### PR TITLE
`Plagiarism checks`: Integrate course wide plagiarism handling with notifications

### PR DESCRIPTION
**Pull request by filip-gregurevic**
This PR allows instructors to handle accepted plagiarism cases in a new course-wide overview. Instructors can provide explanations about the case and notify students. Students can then see the plagiarism case in Artemis. They can then react to the accusation with an explanation in case they think the accusation is wrong.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added `@PreAuthorize` and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This allows instructors to notify students to about potential plagiarism cases. The students can give statements to the plagiarism case and the instructor can decide whether to deny or confirm them.

### Steps for Testing

Prerequisites:

1 Instructor
2 Students
1 Text oder modeling exercise with a detected plagiarism
1a.Log in to Artemis as Instructor
1b. Navigate to Course Detail
1c. Open Plagiarism Cases
1d. Asses the plagiarism case and save instructor statement

2a. Log in to Artemis as Student
2b. Open Notifications
2c. Click on the Plagiarism Notification
2d. Respond to the plagiarism Case

3a. Login in again to Artemis as Instructor
3b. Navigate to Course Detail
3c. Open Plagiarism Cases
3d. Assess the student's response

4a. Log in to Artemis again as Student
4b. Open Notifications
4c. Click on the Plagiarism Notification
4d. Check that the status has been updated by the instructor

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

PlagiarismResource.java: 60% | 87% 
PlagiarismService.java: 37% | 70%
plagiarism-cases.service.ts: 67% | 100%
plagiarism-cases.component.ts: 50% | 77%
plagiarism-cases-review.component.ts: 87% | 95%
plagiarism-cases-list.component.ts: 93% | 100%

(Server test coverage will be improved in future as existing functionality wasn't tested before)

### Screenshots

Student viewing plagiarism case:
Bildschirmfoto 2021-11-14 um 12 39 39
Instructor's view:
Bildschirmfoto 2021-11-14 um 12 39 08
Button for instructor in course detail view to view all cases:
Bildschirmfoto 2021-11-14 um 12 38 52
Instructor view of all cases:
Bildschirmfoto 2021-12-11 um 13 34 36
Notification for final decision:
Bildschirmfoto 2021-11-14 um 12 38 38
Notification about possible case:
Bildschirmfoto 2021-12-11 um 13 33 27
Student's statement:
Bildschirmfoto 2021-12-11 um 13 33 47

### CODE REVIEW

**Comment by Bofan-Zhu** [comment]
Successfully got a notification as a student but could not see the submission with plagiarism of either student. 
image

**Review by Malyuk-A — APPROVED** [review]
(I am not sure If I am allowed to approve this but) We tested this PR in today's testing session on TS3 (TS1 still needs to get fixed due to DB changes) and everything worked as expected.

**Review by b-fein — CHANGES_REQUESTED** [review]
Tested during testing session from a student’s perspective: Got the notification and could submit a comment as intended.

However, there was a scenario that did not work if there are multiple notifications:
1. I click the first notification to go to the page where I can submit a comment.
2. I click on another plagiarism notification:
    - The URL updates to show the different ID of the other case.
    - The actual content does not navigate to the different one, though, and remains on the previous one. only when navigating to a different page first and then clicking on the notification the correct page is shown.

**Review by b-fein — DISMISSED** [review]
According to @filip-gregurevic the issue described in the comment above is due to other existing components. I was promised a follow-up PR that fixes the issues to not extend the already big scope of this PR.
In the meantime this should then not be a big problem, as the likelihood of students having multiple requests for their comments on plagiarism cases should (hopefully) be fairly low.

Therefore, I’m changing the change request to an approval, as the actual functionality worked as described.

**Review by willergeorg — DISMISSED** [review]
Tested during testing session. Functionality worked good.

**Comment by krusche** [comment]
@filip-gregurevic can you post more screenshots in the PR and also add the test coverage?

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java [hunk 1]
```diff
@@ -1,46 +1,238 @@
...
+     * Updates an instructor statement on a plagiarismComparison (for one side).
+     * This process will send a notification to the respective student.
+     * I.e. the instructor sets a personal message to one of the accused students.
+     *
+     * @param comparisonId the id of the PlagiarismComparison
+     * @param studentLogin of one of accused students
+     * @param statement of the instructor directed to one of the accused students
+     * @return the instructor statement (convention)
+     */
+    @PutMapping("plagiarism-comparisons/{comparisonId}/{studentLogin}/instructor-statement")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java:118]: does not follow our REST URL best practices, please improve it:
- start with courses/{courseId}
- studentLogin should not follow immediately after the id, instead you need something in between, such as /cases/

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java [hunk 2]
```diff
@@ -1,46 +1,238 @@
...
+
+    /**
+     * Retrieves the plagiarismComparison specified by its Id. The submissions are anonymized for the student.
+     * StudentIds are replaced with "Your Submission" and "Other Submission" based on the requesting user.
+     *
+     * @param comparisonId the id of the PlagiarismComparison
+     * @return the PlagiarismComparison
+     * @throws AccessForbiddenException if the requesting user is not affected by the plagiarism case.
+     */
+    @GetMapping("plagiarism-comparisons/{comparisonId}")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java:155]: does not follow our REST URL best practices, please improve it:
- please start with courses

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java [hunk 3]
```diff
@@ -1,46 +1,238 @@
...
+
+    /**
+     * Updates a student statement on a plagiarismComparison.
+     * I.e. one of the students that is accused of plagiarising updates/sets the respective/individual response/defence
+     *
+     * @param comparisonId of the comparison
+     * @param statement the students statement
+     * @return the student statement
+     */
+    @PutMapping("plagiarism-comparisons/{comparisonId}/student-statement")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java:178]: does not follow our REST URL best practices, please improve it:
- start with courses

File: src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java [hunk 4]
```diff
@@ -1,46 +1,238 @@
...
+     * Updates the final status of a plagiarism comparison concerning one of both students.
+     * This process will send a notification to the respective student.
+     * I.e. an instructor sends his final verdict/decision
+     *
+     * @param comparisonId of the comparison
+     * @param studentLogin of the student
+     * @param statusDTO is the final status of this plagiarism comparison concerning one of both students
+     * @return the final (updated) status of this plagiarism comparison concerning one of both students
+     */
+    @PutMapping("plagiarism-comparisons/{comparisonId}/{studentLogin}/final-status")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java:209]: does not follow our REST URL best practices, please improve it:
- see comment above

Comment (filip-gregurevic) [reply src/main/java/de/tum/in/www1/artemis/web/rest/PlagiarismResource.java]: All endpoints adapted in fe01c3c

**Review by Malyuk-A — DISMISSED** [review]
Looked at the new code changes, the REST URLs are improved now 👍

**Review by b-fein — APPROVED** [review]
Code changes since last review look good. :+1: 

It looks like the bug I mentioned above should be fixed now as well, I have not tested it, though.

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+    it('should save instructor statement for student a', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.instructorStatement = 'instructor statement text a';
+        saveInstructorStatementStub.mockReturnValue(of({ body: instructorStatementA }));
+        comp.courseId = 1;
+
+        comp.saveInstructorStatement('A', 0);
+
+        expect(comp.plagiarismCase.comparisons[0].instructorStatementA).toEqual(instructorStatementA);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:130]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].instructorStatementA).toBe(instructorStatementA);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+    it('should save instructor statement for student b', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.instructorStatement = 'instructor statement text b';
+        saveInstructorStatementStub.mockReturnValue(of({ body: instructorStatementB }));
+        comp.courseId = 1;
+
+        comp.saveInstructorStatement('B', 0);
+
+        expect(comp.plagiarismCase.comparisons[0].instructorStatementB).toEqual(instructorStatementB);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:142]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].instructorStatementB).toBe(instructorStatementB);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        expect(saveInstructorStatementStub).toHaveBeenCalledWith(1, plagiarismComparison.id, studentLoginB, 'instructor statement text b');
+    });
+
+    it('should update status to confirmed for student a', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.courseId = 1;
+
+        comp.updateStatus(true, 0, studentLoginA);
+
+        expect(comp.plagiarismCase.comparisons[0].statusA).toEqual(PlagiarismStatus.CONFIRMED);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:152]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].statusA).toBe(PlagiarismStatus.CONFIRMED);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        expect(updatePlagiarismComparisonFinalStatusStub).toHaveBeenCalledWith(1, plagiarismComparison.id, true, studentLoginA);
+    });
+
+    it('should update status to denied for student a', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.courseId = 1;
+
+        comp.updateStatus(false, 0, studentLoginA);
+
+        expect(comp.plagiarismCase.comparisons[0].statusA).toEqual(PlagiarismStatus.DENIED);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:162]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].statusA).toBe(PlagiarismStatus.DENIED);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        expect(updatePlagiarismComparisonFinalStatusStub).toHaveBeenCalledWith(1, plagiarismComparison.id, false, studentLoginA);
+    });
+
+    it('should update status to confirmed for student b', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.courseId = 1;
+
+        comp.updateStatus(true, 0, studentLoginB);
+
+        expect(comp.plagiarismCase.comparisons[0].statusB).toEqual(PlagiarismStatus.CONFIRMED);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:172]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].statusB).toBe(PlagiarismStatus.CONFIRMED);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        expect(updatePlagiarismComparisonFinalStatusStub).toHaveBeenCalledWith(1, plagiarismComparison.id, true, studentLoginB);
+    });
+
+    it('should update status to denied for student b', () => {
+        comp.plagiarismCase = plagiarismCase;
+        comp.courseId = 1;
+
+        comp.updateStatus(false, 0, studentLoginB);
+
+        expect(comp.plagiarismCase.comparisons[0].statusB).toEqual(PlagiarismStatus.DENIED);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-list.component.spec.ts:182]:
```suggestion
        expect(comp.plagiarismCase.comparisons[0].statusB).toBe(PlagiarismStatus.DENIED);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        getPlagiarismComparisonForStudentStub.mockReturnValue(of({ body: plagiarismCaseA }));
+
+        comp.ngOnInit();
+        tick();
+
+        expect(comp.comparisonId).toBe(1);
+        expect(getPlagiarismComparisonForStudentStub).toHaveBeenCalledWith(1, 1);
+        expect(comp.exercise).toEqual(textExercise);
+        expect(comp.isStudentA).toBe(true);
+        expect(comp.instructorStatement).toEqual(instructorStatementA);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:99]:
```suggestion
        expect(comp.instructorStatement).toBe(instructorStatementA);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        expect(comp.studentStatement).toEqual(studentStatementA);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:100]:
```suggestion
        expect(comp.studentStatement).toBe(studentStatementA);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+        getPlagiarismComparisonForStudentStub.mockReturnValue(of({ body: plagiarismCaseB }));
+
+        comp.ngOnInit();
+        tick();
+
+        expect(comp.comparisonId).toBe(1);
+        expect(getPlagiarismComparisonForStudentStub).toHaveBeenCalledWith(1, 1);
+        expect(comp.exercise).toEqual(textExercise);
+        expect(comp.isStudentA).toBe(false);
+        expect(comp.instructorStatement).toEqual(instructorStatementB);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:113]:
```suggestion
        expect(comp.instructorStatement).toBe(instructorStatementB);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        expect(comp.studentStatement).toEqual(studentStatementB);
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:114]:
```suggestion
        expect(comp.studentStatement).toBe(studentStatementB);
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+        comp.isStudentA = true;
+        comp.comparison = plagiarismComparisonA;
+        comp.comparisonId = 1;
+        comp.studentStatement = 'statement test';
+        comp.courseId = 1;
+
+        comp.saveStudentStatement();
+        tick();
+
+        expect(comp.comparison.studentStatementA).toEqual('statement test');
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:157]:
```suggestion
        expect(comp.comparison.studentStatementA).toBe('statement test');
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+        comp.isStudentA = false;
+        comp.comparison = plagiarismComparisonB;
+        comp.comparisonId = 2;
+        comp.studentStatement = 'statement test';
+        comp.courseId = 1;
+
+        comp.saveStudentStatement();
+        tick();
+
+        expect(comp.comparison.studentStatementB).toEqual('statement test');
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-cases-review.component.spec.ts:171]:
```suggestion
        expect(comp.comparison.studentStatementB).toBe('statement test');
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts [hunk 17]
```diff
@@ -128,19 +128,19 @@ describe('Plagiarism Inspector Component', () => {
     it('should get the minimumSize tootip', () => {
         comp.exercise = { type: ExerciseType.PROGRAMMING } as Exercise;
 
-        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimum-size-tooltip');
+        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimumSizeTooltip');
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts:131]:
```suggestion
        expect(comp.getMinimumSizeTooltip()).toBe('artemisApp.plagiarism.minimumSizeTooltip');
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts [hunk 18]
```diff
@@ -128,19 +128,19 @@ describe('Plagiarism Inspector Component', () => {
...
     });
 
     it('should get the minimumSize tootip for modeling', () => {
         comp.exercise = { type: ExerciseType.MODELING } as Exercise;
 
-        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimum-size-tooltip-modeling');
+        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimumSizeTooltipModeling');
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts:137]:
```suggestion
        expect(comp.getMinimumSizeTooltip()).toBe('artemisApp.plagiarism.minimumSizeTooltipModeling');
```

File: src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts [hunk 19]
```diff
@@ -128,19 +128,19 @@ describe('Plagiarism Inspector Component', () => {
...
     });
 
     it('should get the minimumSize tootip for text', () => {
         comp.exercise = { type: ExerciseType.TEXT } as Exercise;
 
-        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimum-size-tooltip-text');
+        expect(comp.getMinimumSizeTooltip()).toEqual('artemisApp.plagiarism.minimumSizeTooltipText');
```
Comment (willergeorg) [inline src/test/javascript/spec/component/plagiarism/plagiarism-inspector.component.spec.ts:143]:
```suggestion
        expect(comp.getMinimumSizeTooltip()).toBe('artemisApp.plagiarism.minimumSizeTooltipText');
```

**Review by willergeorg — COMMENTED** [review]
Tested on TS2. Functionality worked fine. I left you some suggestions regarding the client tests.

**Review by Malyuk-A — DISMISSED** [review]
Since my last code review, only minor client test changes were added. Re-approve

**Review by willergeorg — APPROVED** [review]
Thanks for incorporating my suggestions. Re-approve.

**Pull request merged**